app/lib/languages.rb

```diff
@@ -0,0 +1,52 @@
+# frozen_string_literal: true
+
+# Helper methods for languages
+module Languages
+  def self.supported_language?(lang)
+    return false if lang.nil?
+
+    # If the supported field exists at all for a language, then
+    # that language can be considered supported
+    all_languages&.dig(lang.to_sym, :supported)
+  end
+
+  # Even though some languages may be supported, we are unable to send
+  # voice-calls in that language (typically due to Twilio limitations)
+  def self.voice_unsupported?(lang)
+    return false if lang.nil?
+
+    !all_languages&.dig(lang.to_sym, :supported, :phone)
+  end
+
+  def self.all_languages
+    PATIENT_HELPER_FILES[:languages]
+  end
+
+  def self.attempt_language_matching(lang)
+    # This function returns the `lang` passed in if it can't be matched
+    # Else returns the matched 'lang' iso code (stringified)
+    matched_val = normalize_and_get_language_name(lang)
+    matched_val ? matched_val.to_s : lang
+  end
+
+  # This function will attempt to match the input to a language in the system
+  # PARAM: `lang` can be a three-letter iso-639-2t code, a two-letter iso-639-1 code, or the name (not case sensitive)
+  # PARAM EXAMPLES: 'eng', 'en', 'English', 'ENGLISH' <-- All will map to 'eng'
+  # RETURN VALUE: `nil` if unmatchable, else the three-letter iso code ('eng')
+  def self.normalize_and_get_language_name(lang)
```

Comment:
I want to do some benchmarks on this method especially since it runs on every imported patient and we just went through Import performance problems with 1.27

Reply:
This method was changed a bit in https://github.com/SaraAlert/SaraAlert/pull/803/commits/304e2100858d3edf397d89025e3e70de1540df2c to use some new data structures and binary search. The overall performance increase for average case is below:

As you can see, an order of magnitude more performant.
```
Warming up --------------------------------------
             current     1.326k i/100ms
         alpha order   104.073k i/100ms
Calculating -------------------------------------
             current     13.731k (± 3.3%) i/s -     68.952k in   5.027483s
         alpha order      1.039M (± 2.6%) i/s -      5.204M in   5.009701s
```
However, it is worse in the best-case for the previous implementation (Lookup for 'English') since English was the first language in the previous data structure. Shown below:

```
Warming up --------------------------------------
             current   146.811k i/100ms
         alpha order    95.099k i/100ms
Calculating -------------------------------------
         alpha order    952.340k (± 7.4%) i/s -      4.755M in   5.021531s

Comparison:
             current:  1471131.3 i/s
         alpha order:   952339.8 i/s - 1.54x  (± 0.00) slower
```